### PR TITLE
[RyuJIT/ARM32] [ReadyToRun] Fix target register for invocation to Thunk

### PR DESCRIPTION
**Pull request by hseok-oh**
Fix target register for invocation to Thunk code from generated by ReadyToRun
Target register: REG_VIRTUAL_STUB_PARAM (R4, same as LegacyJIT)

related issue: #10654 #8496

### CODE REVIEW

**Comment by hseok-oh** [comment]
## Before

```
IN0024: 000074     mov     r0, r4
IN0025: 000076     mov     r1, r7
IN0026: 000078     mov     r2, r8
IN0027: 00007A     mov     r3, r9
IN0028: 00007C     movw    r12, LOW RELOC 0x7459f93c
IN0029: 000080     movt    r12, HIGH RELOC 0x7459f93c
IN002a: 000084     ldr     r12, [r12]
IN002b: 000088     blx     r12          // System.Int32[,,,]:Set(int,int,int,int,int):this
```

## After

```
IN0025: 000074      mov     r0, r5
IN0026: 000076      mov     r1, r8
IN0027: 000078      mov     r2, r9
IN0028: 00007A      mov     r3, r10
IN0029: 00007C      movw    r12, LOW RELOC 0x7459f93c
IN002a: 000080      movt    r12, HIGH RELOC 0x7459f93c
IN002b: 000084      ldr     r4, [r12]
IN002c: 000088      blx     r4          // System.Int32[,,,]:Set(int,int,int,int,int):this
```

**Comment by hseok-oh** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build please

**Comment by hseok-oh** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please
@dotnet-bot test Windows_NT x64 Debug Build and Test please

**Comment by hseok-oh** [comment]
@BruceForstall Changing `REG_VIRTUAL_STUB_PARAM` may affect ARM32 legacyJIT so it needs more modification & tests. But I agree with your review.

**Comment by hseok-oh** [comment]
@BruceForstall I defined RBM_R2R_INDIRECT_PARAM and merge routine with ARM64. Instead I modified SaveWorker that generates thunk to process changed indir register. PTAL.

File: src/jit/lower.cpp [hunk 1]
```diff
@@ -2542,19 +2542,14 @@ GenTree* Lowering::LowerDirectCall(GenTreeCall* call)
             GenTree* indir    = Ind(cellAddr);
 
 #ifdef FEATURE_READYTORUN_COMPILER
-#if defined(_TARGET_ARM64_)
+#if defined(_TARGET_ARM64_) || defined(_TARGET_ARM_)
             // For arm64, we dispatch code same as VSD using X11 for indirection cell address,
```
Comment (BruceForstall) [inline src/jit/lower.cpp]: Please use `#if defined(_TARGET_ARMARCH_)` instead

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -5524,7 +5524,7 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
             }
 
 #ifdef FEATURE_READYTORUN_COMPILER
-#ifdef _TARGET_ARM64_
+#if defined(_TARGET_ARM64_) || defined(_TARGET_ARM_)
             if (tree->gtCall.IsR2RRelativeIndir())
```
Comment (BruceForstall) [inline src/jit/gentree.cpp]: use `#if defined(_TARGET_ARMARCH_)`

File: src/jit/target.h [hunk 3]
```diff
@@ -1376,6 +1376,11 @@ typedef unsigned short regPairNoSmall; // arm: need 12 bits
   #define RBM_VIRTUAL_STUB_PARAM          RBM_R4
   #define PREDICT_REG_VIRTUAL_STUB_PARAM  PREDICT_REG_R4
 
+  // R2R indirect call. Use the same registers as VSD
+  #define REG_R2R_INDIRECT_PARAM          REG_R4
+  #define RBM_R2R_INDIRECT_PARAM          RBM_R4
+  #define PREDICT_REG_RER_INDIRECT_PARAM  PREDICT_REG_R4
+
```
Comment (BruceForstall) [inline src/jit/target.h]: Don't define `PREDICT_REG_RER_INDIRECT_PARAM` -- it's unused

**Comment by BruceForstall** [comment]
LGTM

@janvorli can you think of any reason changing this register on arm from r12 to r4 wouldn't work?

**Comment by hseok-oh** [comment]
@BruceForstall Fixed it. PTAL

**Pull request merged**